### PR TITLE
Fixes #3150 - BunJS not crashing when using Bun.build and then consume the client.

### LINKED ISSUE #3150: hono/client hc<>() breaks when using with BunJS build
### What version of Hono are you using?

4.5.0

### What runtime/platform is your app running on?

Bun

### What steps can reproduce the bug?

import the hono/client hc to a file and build it using BunJS Bun.build(). Then try to consume it, this will crash the execution.

### What is the expected behavior?

No errors.

### What do you see instead?

BunJS crash.

### Additional information

Bun related issues:
https://github.com/oven-sh/bun/issues/11079
https://github.com/oven-sh/bun/issues/11880

Migrating the client class ClientRequestImpl to a function fixes the issue.

### PR DESCRIPTION
**Pull request by AlbertSabate**
### The author should do the following, if applicable

- [❌] Add tests
- [✔️] Run tests
- [✔️] `bun run format:fix && bun run lint:fix` to format the code
- [❌] Add [TSDoc](https://tsdoc.org/)/[JSDoc](https://jsdoc.app/about-getting-started) to document the code

While we could wait for BunJS to fix it, we could move the client to use functions instead of a class. This should not change the behavior of the client. Also, I took advantage to improve a bit further the types.

### CODE REVIEW

File: src/client/client.ts [hunk 1]
```diff
@@ -183,14 +180,15 @@ export const hc = <T extends Hono<any, any, any>>(
         'ws'
       )
       const targetUrl = new URL(webSocketUrl)
-      for (const key in opts.args[0]?.query) {
-        targetUrl.searchParams.set(key, opts.args[0].query[key])
+      for (const key in opts.args[0]?.query ?? []) {
+        const value = opts.args[0].query[key]
+        targetUrl.searchParams.set(key, Array.isArray(value) ? value.join(',') : value)
```
Comment (AlbertSabate) [inline src/client/client.ts:185]: Please someone to review this line if looks correct.

**Comment by yusukebe** [comment]
@AlbertSabate 

I don't want to change the class style to the function style. So, can you remove the change for class to function, and can you keep only the type and the search params things?

**Comment by AlbertSabate** [comment]
Removed the class change as mentioned. I left only the small improvements I could find. There is more room to improve the types and remove the as UnionToIntersection<Client<T>>, but I think is good enough to leave it like this.

**Comment by yusukebe** [comment]
Hi @AlbertSabate 

This PR breaks the types:

CleanShot 2024-07-20 at 22 36 56@2x

And there are not enough tests. I can't accept this PR. So, I'll close this PR now and create another PR to solve the Search Param's issue with you as a co-author. Thanks.

**Pull request closed without merging**